doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+# The Life of a Connection
+
+This document describes, in broad strokes, how the Conduit proxy works with regard to
+application traffic. We assume that inbound and outbound routing are basically the same,
+modulo a few details that are called out in the proper context.
+
+## Accepting a connection
+
+When an application instance (app) calls a networked service, it resolves the DNS address
+of the service's name to an IP address. It initiates a connection to this address;
+however, because `conduit-proxy-init` has previously configured `iptables(8)`, the
+connection is established to the local `conduit-proxy`.
+
+When the proxy accepts a connection, it first determines the original destination address
+(i.e. the address bound by the client's DNS lookup) via the `SO_ORIGINAL_DST` socket
+option.
+
+It then determines whether there is policy associated with the remote peer's address
+and/or the original destination address. This policy may, for instance, pre-determine how
```

Comment:
@briansmith 

> I don't think it should addresses for this. It should use origins (scheme + host + port) using the information from the TLS SNI extension and/or the Host/:authority HTTP header field.

> I think what you're getting at is that we need to map the target port to the hostname to determine which services the connection was targetted at.

Given a destination address, the proxy at least needs to determine whether the connection should be handled opaquely (i.e. without protocol detection). We can generalize this to say that users should be able to override protocol detection with policy.

This policy must necessarily be determined only by the transport metadata, though I think we could practically constrain this to the original destination address: in the inbound case it is basically only determined by port, though in the outbound case it probably needs to be determined by ip & port).

Once the connection is determined to be opaque--either by policy or by detection--the proxy needs to determine what types of connection-level policies apply to the forwarded tcp stream (e.g. tls, authz, timeouts, lifetimes, buffering requirements, etc). Let's consider this out of scope for the moment.

> There might not be a Kubernetes service object for each service the pod actually handles, e.g. if the pod was addressed by some means other than a service.

First, I think we need to figure out the API the proxy needs to bypass protocol detection. After that, we can draw up a configuration scheme--by port-name convention, Service annotation, CRD, etc--that we can use to drive this API.

Reply:
> First, I think we need to figure out the API the proxy needs to bypass protocol detection.

The first approach would probably be to support two additional environment variables:
* `CONDUIT_PROXY_INBOUND_OPAQUE_PORTS`: a space-separated list of port numbers
* `CONDUIT_PROXY_OUTBOUND_OPAQUE_DESTINATIONS`: a space-separated list of host:port strings--we'll have to do typical service discovery on the host names so that the proxy is dynamic etc.

These environment variables could bet set by `inject` in the short-term and later managed by the control plane (i.e. to support admission control injection).

This means that the opaqueness of endpoints is not dynamically controllable at runtime, which is probably an acceptable tradeoff.

Reply:
> This policy must necessarily be determined only by the transport metadata, though I think we could practically constrain this to the original destination address: in the inbound case it is basically only determined by port, though in the outbound case it probably needs to be determined by ip & port).

> Once the connection is determined to be opaque--either by policy or by detection--the proxy needs to determine what types of connection-level policies apply to the forwarded tcp stream (e.g. tls, authz, timeouts, lifetimes, buffering requirements, etc).

In the case of policies for completely opaque traffic, I agree. However, in the case of HTTP-specific and/or TLS-specific policies, I think we can use more information from inside the connection (e.g. the Host header field or TLS extensions) to handle policy and maybe even to choose which policy to use. It's hard to say until we actually define the policy language.

> Let's consider this out of scope for the moment.

I was just about to write the same thing! We need to write down the specifics of the policies to decide how to handle them.

> There might not be a Kubernetes service object for each service the pod actually handles, e.g. if the pod was addressed by some means other than a service.
> 
> First, I think we need to figure out the API the proxy needs to bypass protocol detection. After that, we can draw up a configuration scheme--by port-name convention, Service annotation, CRD, etc--that we can use to drive this API.

I am doubtful that would work well. I think we need to define the policy mechanism's UI, then its operational aspects (how its stored, edited, and distributed) and then determine how to implement it.

In the short term, we need to decide some things to make progress and I suggest we just to the simplest thing that works given our current feature set, without trying to anticipate what changes will be needed for policy.